File: software-cracking-10weeks/code/week02/die_sim.py
# -*- coding: utf-8 -*-
"""
02_die_sim.py — Mô phỏng Detect It Easy (DIE): nhận diện loại file & dấu hiệu compiler.

Mục tiêu: minh hoạ cách DIE đọc magic bytes và đặc trưng header để đoán
          file là PE/ELF/Mach-O và compiler có thể là gì.
Đầu vào : đường dẫn một file (chỉ ĐỌC, không thực thi).
Đầu ra : loại file, kiến trúc, gợi ý compiler.
An toàn : chỉ đọc vài byte đầu, không chạy binary, không ghi đè file.
"""
import sys

# Magic bytes nhận diện định dạng.
MAGIC = {
    b"MZ": ("PE (Windows)", "x86/x64 — xem thêm PE header"),
    b"\x7fELF": ("ELF (Linux)", "x86/x64/ARM — xem e_machine"),
    b"\xcf\xfa\xed\xfe": ("Mach-O (macOS 32-bit)", "little-endian"),
    b"\xfe\xed\xfa\xcf": ("Mach-O (macOS 32-bit)", "big-endian"),
    b"\xcf\xfa\xed\xfe\x0c": ("Mach-O 64-bit (arm64/x86_64)", "little-endian"),
}


def detect_magic(data: bytes) -> str:
    """Trả về mô tả định dạng dựa trên magic bytes."""
    for magic, desc in MAGIC.items():
        if data.startswith(magic):
            return f"{desc[0]} — {desc[1]}"
    return "Không nhận diện được (file text, script, hoặc định dạng khác)"


def guess_compiler(data: bytes) -> str:
    """Gợi ý compiler từ các chuỗi thường thấy trong binary (chỉ minh hoạ)."""
    # Chỉ quét một phần đầu để tránh đọc toàn bộ file lớn.
    sample = data[:1_000_000]
    hints = [
        (b"MinGW", "MinGW (GCC cho Windows)"),
        (b"GCC:", "GCC"),
        (b"MSVC", "Microsoft Visual C++"),
        (b".NETFramework", ".NET (C#/VB.NET)"),
        (b"UPX", "Có thể bị pack bằng UPX"),
        (b"Delphi", "Delphi/Borland"),
    ]
    found = [name for sig, name in hints if sig in sample]
    return ", ".join(found) if found else "Không rõ (thử DIE thật để xác nhận)"


def main() -> None:
    if len(sys.argv) < 2:
        print("Cách dùng: python die_sim.py <file>")
        sys.exit(1)

    path = sys.argv[1]
    try:
        with open(path, "rb") as f:
            data = f.read()
    except OSError as e:
        print(f"Lỗi đọc file: {e}")
        sys.exit(1)

    print(f"File      : {path}")
    print(f"Kích thước: {len(data)} bytes")
    print(f"Định dạng : {detect_magic(data)}")
    print(f"Compiler  : {guess_compiler(data)}")


if __name__ == "__main__":
    main()

File: software-cracking-10weeks/code/week02/die_sim.cpp
// 02_die_sim.cpp — Mô phỏng Detect It Easy: nhận diện định dạng + gợi ý compiler
// Mục tiêu: đọc magic bytes và đặc trưng header để đoán PE/ELF/Mach-O + compiler.
// An toàn: chỉ ĐỌC đầu file, không thực thi binary.  g++ -std=c++17 -O2 die_sim.cpp -o demo
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstdint>

int main(int argc, char** argv) {
    if (argc < 2) { std::cout << "Cách dùng: ./demo <file>\n"; return 1; }
    std::ifstream f(argv[1], std::ios::binary);
    if (!f) { std::cout << "Không mở được file.\n"; return 1; }
    std::vector<char> data((std::istreambuf_iterator<char>(f)), {});
    std::cout << "File: " << argv[1] << " (" << data.size() << " bytes)\n";

    // 1. Magic bytes.
    auto starts = [&](const char* s){ return data.size() >= 4 && std::string(data.begin(), data.begin()+4) == s; };
    if (starts("MZ"))                    std::cout << "Định dạng: PE (Windows)\n";
    else if (data.size() >= 4 && (unsigned char)data[0] == 0x7F && data[1]=='E' && data[2]=='L' && data[3]=='F')
                                          std::cout << "Định dạng: ELF (Linux)\n";
    else if (starts("\xCF\xFA\xED\xFE"))  std::cout << "Định dạng: Mach-O 32-bit (little-endian)\n";
    else if (starts("\xFE\xED\xFA\xCF"))  std::cout << "Định dạng: Mach-O 32-bit (big-endian)\n";
    else                                 std::cout << "Định dạng: không nhận diện (text/script?)\n";

    // 2. Gợi ý compiler từ chuỗi trong phần đầu.
    std::string s(data.begin(), data.begin() + std::min<size_t>(data.size(), 1 << 20));
    auto has = [&](const char* p){ return s.find(p) != std::string::npos; };
    std::cout << "Compiler : ";
    if (has("MinGW"))          std::cout << "MinGW (GCC)";
    else if (has("GCC:"))      std::cout << "GCC";
    else if (has("MSVC"))      std::cout << "Microsoft Visual C++";
    else if (has(".NETFramework")) std::cout << ".NET (C#/VB.NET)";
    else if (has("UPX"))       std::cout << "có thể bị pack UPX";
    else                       std::cout << "không rõ (dùng DIE thật để xác nhận)";
    std::cout << "\n";
    return 0;
}
